File: Stasis.Game/Scenes/Game/Player/Cursor.cs
using System.Numerics;
using Microsoft.VisualBasic;
using Raylib_cs;
using Stasis.Content.Settings;
using Stasis.Engine;
using Stasis.Engine.GFX;

namespace Stasis.Game.Scenes.Game.Player;

public class Cursor {

    private static float CLAMP_SINGLE = (6.0f - 0.525f) / 2.0f;
    private Vector2 CLAMP = new Vector2(CLAMP_SINGLE, CLAMP_SINGLE);

    public Sprite Sprite;

    public Vector2 Position = Vector2.Zero;

    public Vector2 ClampedPosition = Vector2.Zero;

    public Cursor(Vector3 initialPosition, Vector3 rotation, Vector2 scale, string texPath) {
        Sprite = Sprite.MakePlane(initialPosition, rotation, scale, texPath);
    }

    public void Render() {
        Sprite.Render();
    }

    public void ApplyParallax(ref Camera camera, ref Sprite grid) {
        grid.Position.X = -ClampedPosition.X * (Global.Settings.Camera.GridParallax / 50f);
        grid.Position.Z = -ClampedPosition.Y * (Global.Settings.Camera.GridParallax / 50f);

        camera.Position.X = -ClampedPosition.X * (Global.Settings.Camera.CameraParallax / 50f);
        camera.Position.Y = ClampedPosition.Y * (Global.Settings.Camera.CameraParallax / 50f);
        camera.Target.X = -ClampedPosition.X * (Global.Settings.Camera.CameraParallax / 50f);
        camera.Target.Y = ClampedPosition.Y * (Global.Settings.Camera.CameraParallax / 50f);
    }

    public void ProcessInput() {
        var sensFactor = Global.Settings.Cursor.Sensitivity / 50f;
        var delta = Vector2.One * (InputManager.MouseDelta * sensFactor);
        Position -= delta;

        var gridParallax = Vector2.One * (Global.Settings.Camera.GridParallax / 50f);

        var posClamp = CLAMP - (Position * gridParallax);
        var negClamp = -(CLAMP + (Position * gridParallax));

        ClampedPosition = Vector2.Clamp(Position, negClamp, posClamp);
        if(Global.Settings.Cursor.Clamped) Position = ClampedPosition;
    
        Sprite.Position = new Vector3(Position.X, 0f, Position.Y);
    }
}

File: src/scenes/game/player/cursor.rs
use tracing::info;

use crate::{content::settings::{CameraSettings, CursorSettings}, core::{gfx::{objects::sprite3d::Sprite3dObject, Graphics, ObjectId}, Vector2}};

use super::grid::Grid;

pub struct Cursor {
    id: ObjectId,
    
    scale: f32,
    sensitivity: f32,

    locked_to_grid: bool,

    camera_parallax: f32,
    grid_parallax: f32,

    pub pos: Vector2,
}

const CLAMP: f32 =  (6.0 - 0.525) / 2.0;

impl Cursor {
    pub fn new(gfx: &mut Graphics, settings: &CursorSettings, camera_settings: &CameraSettings) -> Self {
        Self {
            id: gfx.add_sprite(Sprite3dObject::new_plane(-settings.scale / 2., -settings.scale / 2., 0., settings.scale, settings.scale, "Assets/Game/Cursor.png")),
            pos: Vector2::zero(),
            sensitivity: settings.sensitivity / 100.0,
            locked_to_grid: settings.clamped,
            camera_parallax: camera_settings.camera_parallax / 50.0,
            grid_parallax: camera_settings.grid_parallax / 50.0,
            scale: settings.scale,
        }
    }

    pub fn update(&mut self, gfx: &mut Graphics, delta: Vector2) {
        let cursor = &mut gfx.sprite3ds_mut()[self.id];
        
        cursor.x += delta.x * self.sensitivity;
        cursor.y -= delta.y * self.sensitivity;
        

        if self.locked_to_grid {
            cursor.x = cursor.x.clamp(-(CLAMP + (self.pos.x * self.grid_parallax)) - self.scale / 2., CLAMP - (self.pos.x * self.grid_parallax) - self.scale / 2.);
            cursor.y = cursor.y.clamp(-(CLAMP + (self.pos.y * self.grid_parallax)) - self.scale / 2., CLAMP - (self.pos.y * self.grid_parallax) - self.scale / 2.);
        }

        self.pos = Vector2::new(cursor.x - self.scale / 2., cursor.y - self.scale / 2.);

        cursor.update();
    }

    pub fn apply_parallax(&self, gfx: &mut Graphics, grid: &mut Grid) {
        let camera = gfx.bound_camera_mut();
        camera.position.x = -self.pos.x * (self.camera_parallax);
        camera.position.y = -self.pos.y * (self.camera_parallax);

        grid.apply_parallax(gfx, self.pos);

    }
}